common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java

```diff
@@ -192,8 +197,12 @@ private synchronized void initiateRetry() {
   private synchronized boolean shouldRetry(Throwable e) {
     boolean isIOException = e instanceof IOException
       || e.getCause() instanceof IOException;
+    boolean isSaslTimeout = enableSaslRetries &&
+        (e instanceof TransportClient.SaslTimeoutException ||
+            (e.getCause() != null && e.getCause() instanceof TransportClient.SaslTimeoutException));
     boolean hasRemainingRetries = retryCount < maxRetries;
-    return isIOException && hasRemainingRetries && errorHandler.shouldRetryError(e);
+    return (isSaslTimeout || isIOException) &&
+        hasRemainingRetries && errorHandler.shouldRetryError(e);
```

Comment:
If we had sasl failures which triggered a retry - and we finally succeeded - we should reset `retryCount` to `0`.
sasl is part of bootstrap, unlike other failures we retry for.

Reply:
I had to use a map to track the context in which we want to set the retryCount back to 0, since we only want to do it in the case that the exception for that blockId is a SaslTimeoutException. The map is only accessed and mutated within synchronized blocks, so this should allow us to ensure that we are only setting retryCount in the case of SaslTimeoutException retries, and not IOException retries.